types/field_type.go

```diff
@@ -111,13 +111,33 @@ func (ft *FieldType) Init(tp byte) {
 	ft.Decimal = UnspecifiedLength
 }
 
+// GetDefaultFieldLengthAndDecimal returns the default display length (flen)
+// and decimal length for column,  taking into account the sign.
+// Refer: source of MySQL 5.7, adjust_max_effective_column_length()
+func (ft *FieldType) GetDefaultFieldLengthAndDecimal() (flen int, decimal int) {
+	flen, decimal = mysql.GetDefaultFieldLengthAndDecimal(ft.Tp)
+	if !mysql.HasUnsignedFlag(ft.Flag) {
+		return
+	}
+
+	switch ft.Tp {
+	case mysql.TypeInt24, mysql.TypeLong, mysql.TypeTiny, mysql.TypeShort:
+		// Take out the sign
+		flen -= 1
+	case mysql.TypeLonglong:
+		// BINGINT is always 20 no matter the sign
+	}
+
+	return
+}
+
 // CompactStr only considers Tp/CharsetBin/Flen/Deimal.
 // This is used for showing column type in infoschema.
 func (ft *FieldType) CompactStr() string {
 	ts := TypeToStr(ft.Tp, ft.Charset)
 	suffix := ""
 
-	defaultFlen, defaultDecimal := mysql.GetDefaultFieldLengthAndDecimal(ft.Tp)
+	defaultFlen, defaultDecimal := ft.GetDefaultFieldLengthAndDecimal()
```

Comment:
hi @bytewatch , thanks for the contribution!

Would you please add a new function in `field_type_test.go` and add some cases for your fixes?

For some reasons the unit tests were missing, but at least we can add some cases to cover your codes.